url647b55da7c_Nikhilkumar13_emuSr_tgz-pJ8-MainJ8/src/frontend/handlers.java

```diff
@@ -0,0 +1,959 @@
+             FrontEnd.appendToPane(FrontEnd.statuswindow,"Some Error Occured While Opening New File\n",Color.BLACK);
+            //frontend.FrontEnd.statuswindow.setCaretPosition(frontend.FrontEnd.statuswindow.getText().length());
+
+        }
+      }
+
+    void openfile()
+      {
+
+       // FrontEnd.statuswindow.append("Opening File..\n");
+         FrontEnd.appendToPane(FrontEnd.statuswindow,"OpeningFile\n",Color.BLACK);
+        //frontend.FrontEnd.statuswindow.setCaretPosition(frontend.FrontEnd.statuswindow.getText().length());
+        backend.Register.resetRegisters();
+        JFileChooser fileChooser = new JFileChooser();
+        fileChooser.addChoosableFileFilter(new MyCustomFilter());
+        int returnVal = fileChooser.showOpenDialog(new javax.swing.JFrame());
+        if (returnVal == JFileChooser.APPROVE_OPTION)
+        {
+            File file = fileChooser.getSelectedFile();
+            try
+            {
+                // What to do with the file, e.g. display it in a TextArea
+
+                open(file);
+                backend.Register.resetRegisters();
+                clean_branchtable();
+                clean_memtable();
+            } catch (Exception ex)
+            {
+                 FrontEnd.appendToPane(FrontEnd.statuswindow,"Problem In Accessing File"+file.getAbsolutePath()+"\n",Color.BLACK);
+//                FrontEnd.statuswindow.append( + file.getAbsolutePath() + "\n");
+                //frontend.FrontEnd.statuswindow.setCaretPosition(frontend.FrontEnd.statuswindow.getText().length());
+
+            }
+        } 
+        else
+        {
+
+//            FrontEnd.statuswindow.append("File access cancelled by user.\n");
+            FrontEnd.appendToPane(FrontEnd.statuswindow,"File access canceled by user\n",Color.BLACK);
+            //frontend.FrontEnd.statuswindow.setCaretPosition(frontend.FrontEnd.statuswindow.getText().length());
+
+        }    // TODO add your handling code here:
+
+      }
+
+    void savefile()
+      {
+//        FrontEnd.statuswindow.append("Saving File...\n");
+        FrontEnd.appendToPane(FrontEnd.statuswindow,"Saving File...\n",Color.BLACK);
+        //frontend.FrontEnd.statuswindow.setCaretPosition(frontend.FrontEnd.statuswindow.getText().length());
+        int id = FrontEnd.EditorPane.getSelectedIndex();
+        if (FrontEnd.EditorPane.getToolTipTextAt(id) == null)
+        {
+            JFileChooser fileChooser = new JFileChooser();
+            fileChooser.addChoosableFileFilter(new MyCustomFilter());
+            int returnVal = fileChooser.showSaveDialog(new javax.swing.JFrame());
+            if (returnVal == JFileChooser.APPROVE_OPTION)
+            {
+                try
+                {
+
+                    FrontEnd.filepath = fileChooser.getSelectedFile().getPath() + ".s";
+                    File n = new File(FrontEnd.filepath);
+                    BufferedWriter out = new BufferedWriter(new FileWriter(FrontEnd.filepath));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 156, 157. The resource is closed at line 162. There are other execution paths that do not contain closure statements, for example, when JOptionPane.showMessageDialog throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).